### DISCUSSION TITLE
Sharing data between VMs advice

### DISCUSSION THREAD

**Discussion by ghost** [canton7/Stylet#298] [Q&A]
Hello,
I'm working on an app that have  4 viewmodels(and views): RootViewModel (Window), Chart1ViewModel (usercontrol), Chart2ViewModel (usercontrol), TextViewModel (usercontrol).
The 3 usercontrol are hosted in 3 content controls in the main window. 
Whe the app start the main window prompt a OpenFiledialog to load a binary log.
In old winform I put it in a shared array to access it from the 3 usercontrols, but now I'm struggling to make it work in MVVM; 
the 3 VM should display the same data but in different way (pie chart,line chart and text), and if the user change something in the "text visualizzation" the other controls should reflect the change.
I done a lot of search on the web, but finding only  contradictory answers( use instanse like in winforms, not use, use singleton, not use it etc..) Any advice please?

**Comment by canton7** [comment]
It's up to you: there's no right or wrong way.

The way I would do it is to make a class which is responsible for holding the data, and give the 3 VMs references to this (you can either inject an instance, or make it a singleton - your choice). When one of the VMs changes how the data is visualised, they change some property on this class, and this publishes the change to the other VMs (e.g. through an event), so they can update.

**Comment by ghost** [comment]
Thanks, for your advices!
I've declared ad you suggested in another place as this:
 myUC1ViewModel = New UC1ViewModel
        myUC2ViewModel = New UC2ViewModel
        myUC1ViewModel.ConductWith(Me)
 myUC2ViewModel.ConductWith(Me)
And they are initialized correctly.
Now I need to access the viewmodel from another class  
so I tryed that Dim myVM As UC1ViewModel = TryCast(Me.DataContext, UC1ViewModel)
but in debug I get exception because VS say UC1ViewModel is nothing.
Should I instantiate the 3 usercontrols as Singleton, so when I need to refer them Stylet should retrive me the istance?
Is there any example?

**Comment by ghost** [comment]
I'm getting stuck with 2 iusses:
I'm trying to use the EventAggregator as showed [here:](https://github.com/canton7/Stylet/wiki/The-EventAggregator)  (VB.net)
but I get BC30909'message 'cannot expose type' MyEvent 'outside the project using class' UCGraph2dViewModel' at 
```
    Public Sub Handle(message As MyEvent) Implements IHandle(Of MyEvent).Handle

    End Sub
```

Second iusses is that: I'm learning to use DependencyProperties, and getting exception GetValue and SetValue are not defined because missing inherits of DependencyObject, but I can't inherits it in my viewmodel, because it's already inherits screen, the same in the model because is alread inherits PropertyChangedBase.  How can I solve?
Sorry for my bad english and If my questions seem stupid.

**Reply by canton7** [reply]
> but I get BC30909'message 'cannot expose type' MyEvent 'outside the project using class' UCGraph2dViewModel' at

I'm afraid I'm not familiar with VB I'm afraid.

> Second iusses is that: I'm learning to use DependencyProperties, and getting exception GetValue and SetValue are not defined because missing inherits of DependencyObject, but I can't inherits it in my viewmodel, because it's already inherits screen, the same in the model because is alread inherits PropertyChangedBase.

Don't use DependencyProperties on ViewModels. ViewModels use `INotifyPropertyChanged` to publish changes, not DependencyProperties.

Use DependencyProperties with things like UserControls. If you're writing a custom UserControl (which doesn't have a ViewModel), that will probably need DependencyProperties.

**Reply by ghost** [reply]
Thanks

**Comment by ghost** [comment]
I'm drawing a line chart using something like this code: (I only changed the cycle that create the lines to take them from a list)

```
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Security;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualBasic;

public class VisualHost : FrameworkElement
{
    private DrawingVisual Visual = new DrawingVisual();
    private DrawingContext Context;
    private Pen Pen;

    private Random rand = new Random();

    public VisualHost()
    {
        Pen = new Pen(Brushes.Black, 0.25);
        Pen.Freeze();

        AddVisualChild(Visual);
    }

    public void Draw()
    {
        StreamGeometry Geometry = new StreamGeometry();

        using (StreamGeometryContext GeometryContext = Geometry.Open())
        {
            GeometryContext.BeginFigure(new Point(rand.NextDouble() * 500, rand.NextDouble() * 500), false, false);

            for (int i = 1; i <= 1000; i++)

                GeometryContext.LineTo(new Point(rand.NextDouble() * 500, rand.NextDouble() * 500), true, true);
        }

        Geometry.Freeze();

        using (var Context = Visual.RenderOpen())
        {

            Context.DrawGeometry(Nothing, Pen, Geometry);
        }
    }

    protected override int VisualChildrenCount
    {
        get
        {
            return 1;
        }
    }

    protected override Visual GetVisualChild(int index)
    {
        return Visual;
    }
}
```
```
and then I call this every time I need to refresh the chart (every few ms):

public static void RefreshDraw()
{
    VisualHost Host = new VisualHost();
    MainWindow.myGrahp2d.mycanvas.Children.Clear();
    MainWindow.myGrahp2d.mycanvas.Children.Add(Host);
    Host.Draw();
}
```
Now I would to use it in mvvm, but I'm still struggling to making it work: main problem is "RefreshDraw" where  I refere to the usercontrol instance and to mycanvas. How should I do it in mvvm?
